comms/dht/src/network_discovery/mod.rs

```diff
@@ -0,0 +1,762 @@
+}
+
+impl Default for NetworkDiscoveryConfig {
+    fn default() -> Self {
+        Self {
+            enabled: true,
+            min_desired_peers: 50,
+            idle_period: Duration::from_secs(30 * 60),
+            idle_after_num_rounds: 10,
+            on_failure_idle_period: Duration::from_secs(5),
+            max_sync_peers: 5,
+        }
+    }
+}
+
+pub struct DhtNetworkDiscovery {
+    context: NetworkDiscoveryContext,
+    event_tx: broadcast::Sender<Arc<DhtEvent>>,
+    shutdown_signal: ShutdownSignal,
+}
+
+impl DhtNetworkDiscovery {
+    pub fn new(
+        config: DhtConfig,
+        node_identity: Arc<NodeIdentity>,
+        peer_manager: Arc<PeerManager>,
+        connectivity: ConnectivityRequester,
+        event_tx: broadcast::Sender<Arc<DhtEvent>>,
+        shutdown_signal: ShutdownSignal,
+    ) -> Self
+    {
+        Self {
+            context: NetworkDiscoveryContext {
+                config,
+                peer_manager,
+                connectivity,
+                node_identity,
+                num_rounds: Default::default(),
+            },
+            event_tx,
+            shutdown_signal,
+        }
+    }
+
+    async fn get_next_event(&mut self, state: &mut State) -> StateEvent {
+        match state {
+            State::Initializing => Initializing::new(&mut self.context).next_event().await,
+            State::Ready(ready) => ready.next_event().await,
+            State::Discovering(discovering) => discovering.next_event().await,
+            State::Waiting(idling) => idling.next_event().await,
+            _ => StateEvent::Shutdown,
+        }
+    }
+
+    fn transition(&mut self, current_state: State, next_event: StateEvent) -> State {
+        let config = &self.config().network_discovery;
+        debug!(
+            target: LOG_TARGET,
+            "Transition triggered from current state `{}` by event `{}`", current_state, next_event
+        );
+        match (current_state, next_event) {
+            (State::Initializing, StateEvent::Initialized) => {
+                State::Ready(DiscoveryReady::initial(self.context.clone()))
+            },
+            (_, StateEvent::Ready) => State::Ready(DiscoveryReady::initial(self.context.clone())),
+            (State::Ready(_), StateEvent::BeginDiscovery(params)) => {
+                State::Discovering(Discovering::new(params, self.context.clone()))
+            },
+
+            (State::Discovering(_), StateEvent::DiscoveryComplete(stats)) => {
+                if stats.has_new_peers() {
+                    self.publish_event(DhtEvent::NetworkDiscoveryPeersAdded(stats.clone()));
+                }
+
+                State::Ready(DiscoveryReady::new(self.context.clone(), stats))
+            },
+            (State::Ready(_), StateEvent::Idle) => State::Waiting(Waiting::new(config.idle_period)),
+            (_, StateEvent::Shutdown) => State::Shutdown,
+            (_, StateEvent::Errored(err)) => {
+                debug!(
```

Comment:
```suggestion
                error!(
```

Reply:
This is the kind of thing you'd want to set up an alert for in the logs. Error makes it easier to find. Warn is probably also fine

Reply:
Agree - I think initially this branch would execute of no sync peers could be selected, but that had since changed to select other peers and try them.